hubverse_annotator/app.py

```diff
@@ -585,6 +655,78 @@ def load_hubverse_table(hub_file: UploadedFile | None):
     return hub_table
 
 
+def load_observed_data(
+    observed_data_file: UploadedFile | None,
+    expected_schema: dict[str, pl.DataType],
+) -> pl.DataFrame:
+    """
+    Loads and validates the observed data table from a
+    Hubverse formatted file. Returns an empty DataFrame
+    with the given schema if no file is provided.
+    Otherwise, read via `load_hubverse_table`, filter to
+    the latest `as_of` date, then validate against
+    `expected_schema` (stopping on failure).
+
+    Parameters
+    ----------
+    observed_data_file : UploadedFile | None
+        Streamlit-uploaded file containing observed data,
+        or None.
+    expected_schema : dict[str, pl.DataType]
+        Mapping of column names to expected Polars dtypes
+        for observed data.
+
+    Returns
+    -------
+    pl.DataFrame
+        The loaded and validated observed data table, or
+        an empty DataFrame conforming to `expected_schema`
+        if no file was uploaded.
+    """
+    if not observed_data_file:
+        return pl.DataFrame(schema=expected_schema)
+    table = load_hubverse_table(observed_data_file)
+    if "as_of" in table.columns:
+        table = table.filter(pl.col("as_of") == pl.col("as_of").max())
+    validate_schema(table, expected_schema, "Observed Data")
```

Comment:
Validation will fail if `as_of` schema is missing, so swap the order of these lines and remove the `if "as_of"` check.